spec/models/dialog_tab_spec.rb

```diff
@@ -26,4 +26,64 @@
       expect(dialog_tab.dialog_fields).to be_empty
     end
   end
+
+  context '#update_dialog_groups' do
+    before(:each) do
+      @dialog_tab = FactoryGirl.create(:dialog_tab)
+      @dialog_groups = FactoryGirl.create_list(:dialog_group, 2)
+      @dialog_groups.first.dialog_fields << FactoryGirl.create_list(:dialog_field, 1)
+      @dialog_tab.dialog_groups << @dialog_groups
+    end
+
+    it 'deletes a dialog group' do
+      groups = [
+        {
+          'id' => @dialog_groups.first.id,
+          'dialog_fields' => []
+        }
+      ]
+      expect do
+        @dialog_tab.update_dialog_groups(groups)
+      end.to change(@dialog_tab.reload.dialog_groups, :count).by(-1)
+    end
+
+    it 'adds a new dialog group' do
+      groups = [
+        {
+          'id' => @dialog_groups.first.id,
+          'dialog_fields' => []
+        },
+        {
+          'id' => @dialog_groups.last.id,
+          'dialog_fields' => []
+        },
+        {
+          'label' => 'new group',
+          'dialog_fields' => [ { 'name' => 'field', 'label' => 'field' } ]
+        }
+      ]
+
+      expect do
+        @dialog_tab.update_dialog_groups(groups)
+      end.to change(@dialog_tab.reload.dialog_groups, :count).by(1)
+    end
+
+    it 'updates a dialog group' do
+      groups = [
+        {
+          'id' => @dialog_groups.first.id,
+          'label' => 'new label',
+          'dialog_fields' => []
+        },
+        {
+          'id' => @dialog_groups.last.id,
+          'dialog_fields' => []
+        }
+      ]
+
+      expect do
+        @dialog_tab.update_dialog_groups(groups)
+      end.to change(@dialog_groups.first.reload, :label)
+    end
+  end
```

Comment:
Same thing here as for the above, two contexts for the 'id' key being present or not, and possibly consolidating the `groups` local into a let and adjusting things within contexts to give some insight into why differences within that object exist.